cpp/include/cudf/binning/bin.hpp

```diff
@@ -0,0 +1,89 @@
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/detail/bin.hpp>
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/types.hpp>
+
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include <rmm/mr/device/per_device_resource.hpp>
+
+namespace cudf {
+
+/**
+ * @addtogroup binning
+ * @{
+ * @file
+ * @brief APIs for binning values.
+ */
+
+// TODO: Figure out the best way to document this enum. It needs to be in the
+// detail header to support that API, but that won't be publicly documented and
+// a using declaration doesn't show the values of the enum.
+/**
+ * @brief Enum used to define whether or not bins include their boundary points.
+ */
+using inclusive = detail::inclusive;
+
+/**
+ * @brief Labels elements based on membership in the specified bins.
+ *
+ * A bin `i` is defined by `left_edges[i], right_edges[i]`. Whether the edges are inclusive or
+ * not is determined by `left_inclusive` and `right_inclusive`, respectively. If either
+ * `left_edges[i]` or `right_edges[i]` is NULL, no value in `input` can be placed in bin `i`.
+ *
+ * A value `input[j]` belongs to bin `i` if `value[j]` is contained in the range `left_edges[i],
+ * right_edges[i]` (with the specified inclusiveness) and `label[j] == i`. If  `input[j]` does not
+ * belong to any bin, then `label[j]` is NULL.
+ *
+ * Notes:
+ *   - If an empty set of edges is provided, all elements in `input` are labeled NULL.
+ *   - NULL elements in `input` belong to no bin and their corresponding label is NULL.
+ *   - NaN elements in `input` belong to no bin and their corresponding label is NULL.
+ *   - Bins must be provided in monotonically increasing order, otherwise behavior is undefined.
+ *   - If two or more bins overlap, behavior is undefined.
+ *   - If nulls are interspersed throughout the edges rather than being at the beginning or end
+ *     (as set by @p edge_null_precedence), behavior is undefined.
+ *
+ * @throws cudf::logic_error if `input.type() == left_edges.type() == right_edges.type()` is
+ * violated.
+ * @throws cudf::logic_error if `left_edges.size() != right_edges.size()`
+ *
+ * @param input The input elements to label according to the specified bins.
+ * @param left_edges Values of the left edge of each bin.
+ * @param left_inclusive Whether or not the left edge is inclusive.
+ * @param right_edges Value of the right edge of each bin.
+ * @param right_inclusive Whether or not the right edge is inclusive.
+ * @param edge_null_precedence Whether nulls in left and right edges are at the beginning or the
+ * end.
+ * @param mr Device memory resource used to allocate the returned column's device.
+ * @return The integer labels of the elements in `input` according to the specified bins.
+ */
+std::unique_ptr<column> bin(
+  column_view const& input,
+  column_view const& left_edges,
+  inclusive left_inclusive,
+  column_view const& right_edges,
+  inclusive right_inclusive,
```

Comment:
I'm not sure who to ask on this but I think you could just accept a single `inclusive` parameter and instruct the caller that it applies to both edges. Seems odd they would have their nulls sorted differently.

Reply:
I think you may have misunderstood the purpose of the `*_inclusive` parameters. These indicate whether a pair of edges represent an open, closed, or half-open interval (e.g. `(0, 1)`, `(0, 1]`, `[0, 1)`, or `[0, 1]`). They don't have anything to do with nulls. We do need separate parameters for the left and right edges as Keith pointed out [here](https://github.com/rapidsai/cudf/issues/7517#issuecomment-792886583).

The `null_precedence` argument below indicates the null sorting, and as you propose it assumes that both sets of edges have the same null ordering.

Reply:
Yes, I did misunderstand. Sorry about that.